File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:

        self.ans = None
        def dfs(node, p, q):
            if node == None:
                return
            if (node.val >= p.val and node.val <= q.val) or (node.val <= p.val and node.val >= q.val):
                self.ans = node
                return
            if (node.val >= p.val and node.val >= q.val):
                dfs(node.left,p,q)
            else:
                dfs(node.right,p,q)
        
        dfs(root,p,q)
        return self.ans

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func goodNodes(root *TreeNode) int {
    var ans int = 0
    dfs(root,root.Val,&ans)
    return ans
}

func dfs(root *TreeNode, maxVal int, ans *int){
    if root == nil{
        return
    }

    if root.Val >= maxVal{
        *ans = *ans + 1
    }

    dfs(root.Right,max(root.Val,maxVal),ans)
    dfs(root.Left,max(root.Val,maxVal),ans)
    return
}
            

